### PR TITLE
Support PyTorch Dataset for oscar

### PR DESCRIPTION
**Pull request by yuanchongtt**
## What do these changes do?

I have done a demo of pytorch dataset API by inheriting torch.utils.data.Dataset. I also write some testcase in tests subfolder. 

## Related issue number#2224

I have try transfer a torch dataloader object to torch script but I have no idea how to do it. Now I am reading raySGD code to find a way to help mars to run torch or tf code more flexible.

### CODE REVIEW

**Comment by qinxuye** [comment]
Try to use mars dataset in `run_pytorch_script`. This requires passing Mars objects into script.

Before Mars 0.7, we have a concept called `named_tensor` and `named_dataframe`, that is, we can name a tensor or df via `obj.execute(name='sth')`, then in the script users can get it by providing a name like `mt.named_tensor(name='sth')`.

But we don't like the concept because, the named tensor/df has no difference with the vanilla one except adding a name, but we have to add more APIs, and maybe for `md.index` or sth, we need `named_index`, that's a shame. Besides, named object bypassed the lifecycle mechanism in Mars, if an object executed with a name, we don't know when to delete it. However, for vanilla objects, if it's garbage collected, the reference count on the server side will minus 1, and deleted if reference count reaches 0.

In conclusion, we need a new way to pass Mars objects between driver and pytorch script.

I suggest to add the ability to `run_script`, which `run_tensorflow_script` and `run_pytorch_script` rely on, like,

```python
import mars.dataframe as md
from mars.remote import run_script

df = md.read_xxx()
run_script('/to/path/script.py', inputs=[df])
```

In the script.

```python
from mars import get_context

df = get_context().get_inputs()[0]

# other parts
```

The hard part is context can be got inside the script, and `get_inputs()` is absent for [context](https://github.com/mars-project/mars/blob/master/mars/core/context.py).

**Comment by qinxuye** [comment]
#2251 has been merged to master branch, now could leverage it to support PyTorch dataset API.

**Comment by yuanchongtt** [comment]
add Pytorch dataset that inherits from torch.util.data.Dataset and overwrite \_\_len\_\_ \_\_getitem\_\_. It can pass mars.Tensor mars.DataFrame, mars.Seriable, numpy.ndarray, panda.DataFrame, panda.Seriable and return torch Dataset.

File: mars/learn/contrib/pytorch/__init__.py [hunk 1]
```diff
@@ -13,6 +13,8 @@
 # limitations under the License.
 
 from .run_script import run_pytorch_script
+from .dataset import MarsDataset
```
Comment (qinxuye) [inline mars/learn/contrib/pytorch/__init__.py:16]: Add `# noqa: F401  # pylint: disable=unused-import` to end of this line to pass codacy.

File: mars/learn/contrib/pytorch/__init__.py [hunk 2]
```diff
@@ -13,6 +13,8 @@
...
+from .sampler import SequentialSampler, RandomSampler, SubsetRandomSampler
```
Comment (qinxuye) [inline mars/learn/contrib/pytorch/__init__.py:17]: Ditto.

File: mars/learn/contrib/pytorch/dataset.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    def __len__(self):
+        return self._tileables[0].shape[0]
+
+    def __getitem__(self, index):
+        return tuple(self.get_data(t, index) for t in self._tileables)
+
+    @staticmethod
+    def get_data(t, index):
+        if isinstance(t, TENSOR_TYPE):
```
Comment (qinxuye) [inline mars/learn/contrib/pytorch/dataset.py:50]: If encounter Mars object, shall us `execute` it for the first time. Then, we can do index fetch only.

This is based on the fact that.

```python
df.execute()
df.iloc[:5].fetch()  # it's allowed to fetch subset of executed objects, but it would be very quick.
df.iloc[5:10].fetch()
```

File: mars/learn/contrib/pytorch/tests/test_dataset.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    # Mars tensor
+    data = mt.random.rand(1000, 32, dtype='f4')
+    labels = mt.random.randint(0, 2, (1000, 10), dtype='f4')
+
+    data_verify = data[0].execute().fetch()
+    labels_verify = labels[0].execute().fetch()
+
+    train_dataset = MarsDataset(data, labels)
+    assert isinstance(train_dataset, Dataset)
+    assert (train_dataset[0][0] == data_verify).all()
```
Comment (qinxuye) [inline mars/learn/contrib/pytorch/tests/test_dataset.py:40]: To test equality of Mars tensor and ndarray. please use

```python
np.testing.assert_array_equal(arr1, arr2)
```

**Review by qinxuye — COMMENTED** [review]
LGTM overall. Could you please add a doc to https://github.com/mars-project/mars/tree/master/docs/source/user_guide/learn to tell the users how to integrate with pytorch.

**Comment by qinxuye** [comment]
I didn't find the test that integrates MarsDataset with `run_pytorch_script`, could please add a test as well?

File: mars/learn/contrib/pytorch/dataset.py (hunk 3 above)
Comment (yuanchongtt) [reply mars/learn/contrib/pytorch/dataset.py:50]: That is great advice. I have considered this situation, but forget to optimize it. I will modify it as soon as possible.

File: mars/learn/contrib/pytorch/tests/test_dataset.py (hunk 4 above)
Comment (yuanchongtt) [reply mars/learn/contrib/pytorch/tests/test_dataset.py:40]: That's great advice!

File: mars/learn/contrib/pytorch/dataset.py [hunk 5]
```diff
@@ -48,14 +59,14 @@ def __getitem__(self, index):
     @staticmethod
     def get_data(t, index):
         if isinstance(t, TENSOR_TYPE):
-            return t[index].execute().fetch()
+            return t[index].fetch()
         elif isinstance(t, np.ndarray):
             return t[index]
         elif isinstance(t, DATAFRAME_TYPE):
-            return t.iloc[index].execute().fetch().values
+            return t.iloc[index:index+1].fetch().values[0]
```
Comment (yuanchongtt) [inline mars/learn/contrib/pytorch/dataset.py:66]: The original version is `return t.iloc[index].fetch().values`. It will get a Series data and fetch() to get np.ndarray data. But it got an error `AssertionError: series name in metadata None is not equal to real name 1`. The details list below. Now this version can work well. 

dataset.py:57: in __getitem__
    return tuple(self.get_data(t, index) for t in self._tileables)
dataset.py:57: in <genexpr>
    return tuple(self.get_data(t, index) for t in self._tileables)
dataset.py:69: in get_data
    print(type(t.iloc[index].fetch()))
../../../dataframe/core.py:525: in fetch
    return self._fetch(session=session, **kw)
../../../core/entity/executable.py:122: in _fetch
    return fetch(self, session=session, **kw)
../../../deploy/oscar/session.py:1393: in fetch
    return session.fetch(tileable, *tileables, **kwargs)
../../../deploy/oscar/session.py:1242: in fetch
    return asyncio.run_coroutine_threadsafe(coro, self._loop).result()
/home/yuanchong/.conda/envs/mars-dev/lib/python3.7/concurrent/futures/_base.py:435: in result
    return self.__get_result()
/home/yuanchong/.conda/envs/mars-dev/lib/python3.7/concurrent/futures/_base.py:384: in __get_result
    raise self._exception
../../../deploy/oscar/session.py:1377: in _fetch
    data = await session.fetch(tileable, *tileables, **kwargs)
../../../deploy/oscar/tests/session.py:80: in fetch
    results = await super().fetch(*tileables)
../../../deploy/oscar/session.py:866: in fetch
    result.append(self._process_result(tileable, merged))
../../../deploy/oscar/tests/session.py:69: in _process_result
    self.assert_object_consistent(tileable, result)
../../../tests/core.py:357: in assert_object_consistent
    self.assert_series_consistent(expected, real)

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <mars.deploy.oscar.tests.session.CheckedSession object at 0x7f284544fb50>, expected = Series(op=DataFrameIlocGetItem)
real = 0     0.206287
1     0.881219
2     0.354464
3     0.832302
4     0.316828
5     0.205301
6     0.778543
7     0.42978...83043
26    0.826312
27    0.575012
28    0.435271
29    0.059423
30    0.026819
31    0.124065
Name: 1, dtype: float64

    def assert_series_consistent(self, expected, real):
        series_types = (pd.Series,)
        if cudf is not None:
            series_types += (cudf.Series,)
    
        if not isinstance(real, series_types):
            raise AssertionError(f'Type of real value ({type(real)}) not Series')
        self.assert_shape_consistent(expected.shape, real.shape)
    
        if self._check_options['check_series_name'] and expected.name != real.name:
>           raise AssertionError(f'series name in metadata {expected.name} '
>                                          f'is not equal to real name {real.name}')
>           E           AssertionError: series name in metadata None is not equal to real name 1

**Comment by yuanchongtt** [comment]
In pytest, it seems that we just verity `run_script` return `status == ok`, but I can't verify if the script(pytorch_dataset.py)  is executed correctly. I don't know how to do it with pytest.

**Comment by qinxuye** [comment]
> In pytest, it seems that we just verity `run_script` return `status == ok`, but I can't verify if the script(pytorch_dataset.py) is executed correctly. I don't know how to do it with pytest.

Just do assert in the script.

**Comment by hekaisheng** [comment]
> In pytest, it seems that we just verity `run_script` return `status == ok`, but I can't verify if the script(pytorch_dataset.py) is executed correctly. I don't know how to do it with pytest.

I have pushed a commit to your branch, it should be fixed now.

**Comment by yuanchongtt** [comment]
I have wrote a doc about pytorch and added a test that integrates MarsDataset with run_pytorch_script. But when I pass a mars data to script, the data must `execute`, if not, it raised ` IndexError: Cannot choose from an empty sequence`. And it just occurred in `run_pytorch_script` and `run_tensorflow_script`,  `mars.remote.run_srcipt` didn't raise this error.


========================================================== FAILURES ===========================================================
___________________________________________________ test_MarsDataset_script ___________________________________________________

setup_cluster = <mars.deploy.oscar.session.SyncSession object at 0x7f0ba00428d0>

    @pytest.mark.skipif(not torch_installed, reason='pytorch not installed')
    def test_MarsDataset_script(setup_cluster):
        sess = setup_cluster
        path = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                            'pytorch_dataset.py')

        data = mt.random.rand(1000, 32, dtype='f4')
        labels = mt.random.randint(0, 2, (1000, 10), dtype='f4')

        # data.execute()
        # labels.execute()

>       assert run_pytorch_script(
            path, n_workers=2, data={'feature_data': data, 'labels': labels},
            command_argv=['multiple'], port=9945, session=sess).fetch()['status'] == 'ok'

test_dataset.py:253:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../run_script.py:143: in run_pytorch_script
    return op(inputs).execute(session=session, **(run_kwargs or {}))
../../../../core/entity/executable.py:104: in execute
    return execute(self, session=session, **kw)
../../../../deploy/oscar/session.py:1372: in execute
    **kwargs)
../../../../deploy/oscar/session.py:1214: in execute
    timeout=self._isolated_session.timeout)
/home/yuanchong/.conda/envs/mars-dev/lib/python3.7/concurrent/futures/_base.py:435: in result
    return self.__get_result()
/home/yuanchong/.conda/envs/mars-dev/lib/python3.7/concurrent/futures/_base.py:384: in __get_result
    raise self._exception
../../../../deploy/oscar/session.py:1351: in _execute
    await execution_info
/home/yuanchong/.conda/envs/mars-dev/lib/python3.7/asyncio/tasks.py:630: in _wrap_awaitable
    return (yield from awaitable.__await__())
../../../../deploy/oscar/session.py:77: in wait
    return await self._aio_task
../../../../deploy/oscar/session.py:691: in _run_in_background
    raise task_result.error.with_traceback(task_result.traceback)
../../../../services/task/supervisor/processor.py:46: in inner
    return await func(processor, *args, **kwargs)
../../../../services/task/supervisor/processor.py:267: in schedule
    await stage_processor.run()
../../../../services/task/supervisor/stage.py:157: in run
    await self._schedule_subtasks(indep_subtasks)
../../../../services/task/supervisor/stage.py:81: in _schedule_subtasks
    subtasks, [subtask.priority for subtask in subtasks])
../../../../services/scheduling/api.py:67: in add_subtasks
    await self._manager_ref.add_subtasks(subtasks, priorities)
../../../../oscar/backends/context.py:154: in send
    return self._process_result_message(result)
../../../../oscar/backends/context.py:59: in _process_result_message
    raise message.error.with_traceback(message.traceback)
../../../../oscar/backends/pool.py:496: in send
    result = await future
../../../../oscar/api.py:118: in __on_receive__
    return await super().__on_receive__(message)
mars/oscar/core.pyx:348: in __on_receive__
    ???
mars/oscar/core.pyx:320: in mars.oscar.core._BaseActor.__on_receive__
    ???
mars/oscar/core.pyx:323: in mars.oscar.core._BaseActor.__on_receive__
    ???
mars/oscar/core.pyx:326: in mars.oscar.core._BaseActor.__on_receive__
    ???
../../../../services/scheduling/supervisor/manager.py:81: in add_subtasks
    [subtask for subtask in subtasks if not subtask.virtual], priorities)
../../../../oscar/backends/context.py:154: in send
    return self._process_result_message(result)
../../../../oscar/backends/context.py:59: in _process_result_message
    raise message.error.with_traceback(message.traceback)
../../../../oscar/backends/pool.py:496: in send
    result = await future
../../../../oscar/api.py:118: in __on_receive__
    return await super().__on_receive__(message)
mars/oscar/core.pyx:348: in __on_receive__
    ???
mars/oscar/core.pyx:320: in mars.oscar.core._BaseActor.__on_receive__
    ???
mars/oscar/core.pyx:323: in mars.oscar.core._BaseActor.__on_receive__
    ???
mars/oscar/core.pyx:326: in mars.oscar.core._BaseActor.__on_receive__
    ???
../../../../services/scheduling/supervisor/queueing.py:121: in add_subtasks
    bands = await self._assigner_ref.assign_subtasks(subtasks)
../../../../oscar/backends/context.py:154: in send
    return self._process_result_message(result)
../../../../oscar/backends/context.py:59: in _process_result_message
    raise message.error.with_traceback(message.traceback)
../../../../oscar/backends/pool.py:496: in send
    result = await future
../../../../oscar/api.py:118: in __on_receive__
    return await super().__on_receive__(message)
mars/oscar/core.pyx:348: in __on_receive__
    ???
mars/oscar/core.pyx:320: in mars.oscar.core._BaseActor.__on_receive__
    ???
mars/oscar/core.pyx:323: in mars.oscar.core._BaseActor.__on_receive__
    ???
mars/oscar/core.pyx:326: in mars.oscar.core._BaseActor.__on_receive__
    ???
../../../../services/scheduling/supervisor/assigner.py:156: in assign_subtasks
    assigns.append(random.choice(bands))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <random.Random object at 0x55b5861ad2a0>, seq = []

    def choice(self, seq):
        """Choose a random element from a non-empty sequence."""
        try:
            i = self._randbelow(len(seq))
        except ValueError:
>           raise IndexError('Cannot choose from an empty sequence') from None
E           IndexError: Cannot choose from an empty sequence

/home/yuanchong/.conda/envs/mars-dev/lib/python3.7/random.py:261: IndexError

File: mars/learn/contrib/pytorch/dataset.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    DataFrame, Series. Additionally, it's constructor can receive
+    np.ndarray, pd.DataFrame, pd.Series type.
+    """
+    def __init__(self, *tileables, fetch_kwargs=None):
+
+        self._context = get_context()
+        self._tileables = tileables
+        self._fetch_kwargs = fetch_kwargs or dict()
+
+        self._check_and_execute()
```
Comment (qinxuye) [inline mars/learn/contrib/pytorch/dataset.py:45]: I don't think it's a good idea to trigger execution when intializing a dataset, but instead, we can check in `__getitem__`. e.g.

```python
class MarsDataset(Dataset):
    def __init__(self, xxx):
        xxx
        self._executed = False

    def __getitem__(self, index):
        if not self._executed:
            self._check_and_execute()
```

File: mars/learn/contrib/pytorch/dataset.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+    def _check_and_execute(self):
+        for t in self._tileables:
+            if isinstance(t, TENSOR_TYPE):
+                t.execute()
```
Comment (qinxuye) [inline mars/learn/contrib/pytorch/dataset.py:50]: `execute` on each tileable is not so efficient, executing them all together is a better solution.

```python
import mars

mars.execute([t for t in self._tileables if isinstance(t, ENTITY_TYPE)])
```

File: mars/learn/contrib/pytorch/dataset.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            elif isinstance(t, DATAFRAME_TYPE):
+                t.execute()
+            elif isinstance(t, SERIES_TYPE):
+                t.execute()
```
Comment (qinxuye) [inline mars/learn/contrib/pytorch/dataset.py:54]: Add an `else` to raise TypeError if uknown types show up.

File: mars/learn/contrib/pytorch/sampler.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+    Args:
+        data_source (Dataset): dataset to sample from
+    """
+    data_source: Sized
+
+    def __init__(self, data_source):
+        self.data_source = data_source
+
+    def __iter__(self) -> Iterator[int]:
+        return iter(range(len(self.data_source)))
```
Comment (qinxuye) [inline mars/learn/contrib/pytorch/sampler.py:41]: Is it possible to sample a batch data instead of 1 record each time?

File: mars/learn/contrib/pytorch/sampler.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+        assert len(indices) == self.total_size
+
+        # subsample
+        indices = indices[self.rank:self.total_size:self.num_replicas]
+        assert len(indices) == self.num_samples
+
+        return indices
+
+    def __iter__(self):
+        return iter(self.generate_indices())
```
Comment (qinxuye) [inline mars/learn/contrib/pytorch/sampler.py:242]: Maybe can add some prefetch logic in this sampler to accelerate data loader.

File: mars/remote/tests/sample_script.py [hunk 11]
```diff
@@ -13,4 +13,12 @@
...
 import os
-assert os.environ['WORLD_SIZE'] == '2'
+
+
+def main():
+    pass
+
+
+if __name__ == "__main__":
+    assert os.environ['WORLD_SIZE'] == '2'
```
Comment (qinxuye) [inline mars/remote/tests/sample_script.py:23]: This file seems do nothing?

**Review by qinxuye — COMMENTED** [review]
I left some comments.

File: mars/learn/contrib/pytorch/dataset.py (hunk 6 above)
Comment (yuanchongtt) [reply mars/learn/contrib/pytorch/dataset.py]: OK, I have made some modifications, please check it.

File: mars/learn/contrib/pytorch/dataset.py (hunk 7 above)
Comment (yuanchongtt) [reply mars/learn/contrib/pytorch/dataset.py]: Ditto.

File: mars/learn/contrib/pytorch/dataset.py (hunk 8 above)
Comment (yuanchongtt) [reply mars/learn/contrib/pytorch/dataset.py]: OK, I have add a type_check. Please check it!

File: mars/learn/contrib/pytorch/sampler.py (hunk 9 above)
Comment (yuanchongtt) [reply mars/learn/contrib/pytorch/sampler.py:41]: Batch related settings are under torch.utils.data.DataLoader. To avoid making a conflict with torch.utils.data.DataLoader, I didn't consider the batch setting in sampler.

File: mars/learn/contrib/pytorch/sampler.py (hunk 10 above)
Comment (yuanchongtt) [reply mars/learn/contrib/pytorch/sampler.py:242]: I have read the code of mars 0.6. In my opinion, If we do prefetch according to indices. The dataset indices will be reset. So when we call torch.utils.data.DataLoader, the sampler sequence will be wrong. So I didn't add prefetch operation. Could you please give me some specific advice?

File: mars/remote/tests/sample_script.py (hunk 11 above)
Comment (yuanchongtt) [reply mars/remote/tests/sample_script.py]: Yes, this file do nothing, I have reverted it.

**Review by qinxuye — DISMISSED** [review]

**Review by qinxuye — APPROVED** [review]
LGTM

**Review by hekaisheng — APPROVED** [review]
LGTM

**Comment by qinxuye** [comment]
Congrats that this PR got merged, look forward to seeing more contributions from you.
